### PR TITLE
WIP: Introduce nix flake (Part-1)

### PR DESCRIPTION
**Pull request by sbh69840**
This PR will add nix flakes to the todo-app:
Refer to TODO section of README.md to see the tasks for this PR

### CODE REVIEW

File: tutorial/nixify/1_nix_flake.md [hunk 1]
```diff
@@ -3,8 +3,249 @@
...
 
-The series is designed in a way that each blog will work on the previous blog to improve the development experience. First blog will introduce you to the pain points of using a haskell project without Nix, then we will bring in Nix flakes. The Haskell project of choice here is, [todo-app](https://github.com/juspay/todo-app). This app was chosen for two main reasonse: 1) Keep it simple 2) 
+The series is designed in a way that each blog will work on the previous blog to improve the development experience. First blog will introduce you to the pain points of using a haskell project without Nix, then we will bring in Nix flakes. The Haskell project of choice here is, [todo-app](https://github.com/juspay/todo-app/tree/tutorial/part-0). This app was chosen for two main reasons: 1) Keep things simple 2) Demonstrate how to manage dependencies like postgres and postgREST without having to set them up manually.
```
Comment (srid) [inline tutorial/nixify/1_nix_flake.md:6]:
```suggestion
The series is designed in a way that each blog post will work on the previous blog post to improve the development experience. First blog post will introduce you to the pain points of using a haskell project without Nix, then we will bring in Nix flakes. The Haskell project of choice here is, [todo-app](https://github.com/juspay/todo-app/tree/tutorial/part-0). This app was chosen for two main reasons: 1) Keep things simple 2) Demonstrate how to manage runtime dependencies like postgres and postgREST without having to set them up manually.
```

File: tutorial/nixify/1_nix_flake.md [hunk 2]
```diff
@@ -3,8 +3,249 @@
 <!-- This is the first part of the series that will help you improve the dev workflow of your haskell project in every part.  -->
 This is the first part of a series of blogs, with the end-goal of having a better dev workflow for your haskell project. We are aiming for a "works out-of-the-box" experience. [Nix](https://nixos.org) will be used during this process. [Read here](https://haskell.flake.page/nix-rapid) for a rapid introduction to Nix. Throughout this series we assume the reader to have an understanding of Nix and [Nix flakes](https://nixos.wiki/wiki/Flakes).\
```
Comment (srid) [inline tutorial/nixify/1_nix_flake.md:4]:
```suggestion
This is the first part of a series of blog posts, with the end-goal of having a better dev workflow for your haskell project. We are aiming for a "works out-of-the-box" experience. [Nix](https://nixos.org) will be used during this process. [Read here](https://haskell.flake.page/nix-rapid) for a rapid introduction to Nix. Throughout this series we assume the reader to have an understanding of Nix and [Nix flakes](https://nixos.wiki/wiki/Flakes).\
```

File: tutorial/nixify/1_nix_flake.md [hunk 3]
```diff
@@ -3,8 +3,249 @@
...
+- Postgres and PostgREST are installed globally: This will create problems when you want to work on multiple different versions of these dependencies.
+- If the project doesn't specify the versions of the dependencies used, then you might end up in what's called a "Dependency hell".
+- Developer time spent setting up all of those dependencies, when all they want is to quickly make changes and test them.
+\
+How is Nix going to solve the above issues?
+- Postgres server will run using the data directory in the current project directory.
+- Both Postgres and PostgREST are only present in the Nix shell and since they are installed in Read-only filesystem `/nix`, having multiple different versions will not cause any issues as there is no chance of any rewrites by mistake.
+- The `flake.lock` file will contain the locked content hash, revision and source of "inputs" which will ensure the same source is used on further builds.
+\
+All these benefits are listed as per the Nixification that will be happening in this blog. More benefits will be depmonstrated in upcoming blogs.
```
Comment (srid) [inline tutorial/nixify/1_nix_flake.md:31]:
```suggestion
All these benefits are listed as per the Nixification that will be happening in this blog post. More benefits will be depmonstrated in upcoming blog posts.
```

File: tutorial/nixify/1_nix_flake.md [hunk 4]
```diff
@@ -3,8 +3,249 @@
...
+The steps mentioned above are brief, refer to the [README](https://github.com/juspay/todo-app/blob/tutorial/part-0/README.md) for details.\
+These are some of the problems with the steps mentioned above:
+- Postgres and PostgREST are installed globally: This will create problems when you want to work on multiple different versions of these dependencies.
+- If the project doesn't specify the versions of the dependencies used, then you might end up in what's called a "Dependency hell".
+- Developer time spent setting up all of those dependencies, when all they want is to quickly make changes and test them.
+\
+How is Nix going to solve the above issues?
+- Postgres server will run using the data directory in the current project directory.
+- Both Postgres and PostgREST are only present in the Nix shell and since they are installed in Read-only filesystem `/nix`, having multiple different versions will not cause any issues as there is no chance of any rewrites by mistake.
+- The `flake.lock` file will contain the locked content hash, revision and source of "inputs" which will ensure the same source is used on further builds.
```
Comment (srid) [inline tutorial/nixify/1_nix_flake.md]: We could improve these points a bit. Let me think about this.

File: tutorial/nixify/1_nix_flake.md [hunk 5]
```diff
@@ -3,8 +3,249 @@
...
+    
+  };
+}
+
+```
+
+```nix
+{
+  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+  outputs = {self, ...}@inputs: {
```
Comment (srid) [inline tutorial/nixify/1_nix_flake.md:60]:
```suggestion
  outputs = inputs@{self, ...}: {
```

File: tutorial/nixify/1_nix_flake.md [hunk 6]
```diff
@@ -3,8 +3,249 @@
...
+  # Inside outputs
+  apps = forAllSystems ( system: {
+    postgres =
+      let
+        pgsql = pkgs.${system}.postgresql;
+      in
+      {
+        type = "app";
+        program = toString ( 
+          pkgs.${system}.writeShellScript "pg" ''
```
Comment (srid) [inline tutorial/nixify/1_nix_flake.md:190]: I'd use `writeShellApplication` instead as it uses shellcheck automatically.

File: tutorial/nixify/1_nix_flake.md [hunk 7]
```diff
@@ -3,8 +3,249 @@
...
+    };
+  });
+}
+```
+
+## Before vs After
+
+- Installing postgres, setting up server and loading db dump replaced with -> `nix run .#postgres`
+- Installing postgREST webserver, running it replaced with -> `nix run .#postgrest`
+- Installing cabal-install, building the project replaced with -> `nix develop` and `nix build`
```
Comment (srid) [inline tutorial/nixify/1_nix_flake.md:253]: Also `nix run`

**Review by srid — COMMENTED** [review]
Just some initial feedback.

File: tutorial/nixify/1_nix_flake.md (hunk 6 above)
Comment (sbh69840) [reply tutorial/nixify/1_nix_flake.md:190]: I have changed it to `writeShellApplication`! shellcheck is great. I have also added more checks in the script in scenarios where the database is running and data directory already exists.

**Pull request closed without merging**